### PR TITLE
Add io.parmentier.flatpak.streamdeck.

### PR DESCRIPTION
**Pull request by lpyparmentier**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link: https://github.com/timothycrosley/streamdeck-ui/issues/168**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

**Note**: Concerning the minimal set of permission, I'm using `device=all`, but maybe it is possible to do something better ? like `device=udev` and specify a list of devices ? 

**Questions**: the application have access to some files/directory under $HOME, I don't see how since I do not give access any filesystem access (verified on flatseal). And it is capable to use them (when I put any jpg/png on the stream deck, it is working).

### CODE REVIEW

File: io.parmentier.flatpak.streamdeck.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+app-id: io.parmentier.flatpak.streamdeck
+base-version: '5.15'
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:2]:
```suggestion
```

File: io.parmentier.flatpak.streamdeck.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+runtime: org.kde.Platform
+runtime-version: '5.15'
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:4]:
```suggestion
runtime-version: '5.15-21.08'
```

File: flathub.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64"]
```
Comment (hfiguiere) [inline flathub.json:2]: why?

File: io.parmentier.flatpak.streamdeck.yml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+          - mkdir -p /app/include/qt5tmp && cp -R /usr/include/Qt* /app/include/qt5tmp # https://bugreports.qt.io/browse/PYSIDE-787
+          - sed -i 's|\(--include-paths=\)|\1/app/include/qt5tmp:|' sources/pyside2/cmake/Macros/PySideModules.cmake
+
+  - name: streamdeck-ui
+    buildsystem: simple
+    build-commands:
+      - pip3 install --prefix=/app .
+    build-options:
+      build-args:
+        - --share=network
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:63]: this won't work, there is no network access during build. You have to add the other python dependencies as needed.

File: modules/hidapi.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (hfiguiere) [inline modules/hidapi.json:1]: this file isn't used.

File: modules/udev.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+    {
```
Comment (hfiguiere) [inline modules/udev.json:1]: nor is this file

File: io.parmentier.flatpak.streamdeck.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
+app-id: io.parmentier.flatpak.streamdeck
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:1]: give the URL of the app upstream is https://timothycrosley.github.io/streamdeck-ui/ I don't think this app-id is appropriate.

File: io.parmentier.flatpak.streamdeck.appdata.xml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+  <id>io.parmentier.flatpak.streamdeck</id>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>MIT</project_license>
+  <launchable type="desktop-id">io.parmentier.flatpak.streamdeck</launchable>
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.appdata.xml:6]:
```suggestion
  <launchable type="desktop-id">io.parmentier.flatpak.streamdeck.desktop</launchable>
```

(it needs to have `.destkop` added to the actual app-id)

File: io.parmentier.flatpak.streamdeck.yml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+base-version: '5.15'
+runtime: org.kde.Platform
+runtime-version: '5.15'
+sdk: org.kde.Sdk
+command: streamdeck
+finish-ars:
+  - --socket=x11
+  - --socket=session-bus
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:9]: no. you need to determine which dbus name is used and add them here.

File: io.parmentier.flatpak.streamdeck.yml [hunk 10]
```diff
@@ -0,0 +1,67 @@
+app-id: io.parmentier.flatpak.streamdeck
+base-version: '5.15'
+runtime: org.kde.Platform
+runtime-version: '5.15'
+sdk: org.kde.Sdk
+command: streamdeck
+finish-ars:
+  - --socket=x11
```
Comment (hfiguiere) [inline io.parmentier.flatpak.streamdeck.yml:8]:
```suggestion
  - --socket=x11
  - --share=ipc
```

(needed for X11)

File: io.parmentier.flatpak.streamdeck.yml (hunk 9 above)
Comment (lpyparmentier) [inline io.parmentier.flatpak.streamdeck.yml:9]: if not present, it does not appear on the task manager/bar.

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,8 @@
+sdk-dep:
+	flatpak install flathub org.kde.Sdk//5.15
```
Comment (hfiguiere) [inline Makefile:2]:
```suggestion
	flatpak install flathub org.kde.Sdk//5.15-21.08
```

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: io.parmentier.flatpak.streamdeck.yml (hunk 9 above)
Comment (hfiguiere) [reply io.parmentier.flatpak.streamdeck.yml:9]: pass `--log-session-bus` to `flatpak run` to determine what it needs.

**Comment by hfiguiere** [comment]
bot, build io.parmentier.flatpak.streamdeck

**Comment by hfiguiere** [comment]
> > Build [67546 failed](https://flathub.org/builds/#/builders/32/builds/67546)
> 
> Ok so it seems to failed due to the presence of `build-options/build-args`, however I can't build it locally without it (see previous thread).

which previous thread.

**Comment by lpyparmentier** [comment]
> > > Build [67546 failed](https://flathub.org/builds/#/builders/32/builds/67546)
> > 
> > 
> > Ok so it seems to failed due to the presence of `build-options/build-args`, however I can't build it locally without it (see previous thread).
> 
> which previous thread.

My answer to this comment https://github.com/flathub/flathub/pull/2648/files#r752235790

**Comment by hfiguiere** [comment]
> My answer to this comment https://github.com/flathub/flathub/pull/2648/files#r752235790

I don't see anything.

Anyway, there is no network access during build, but since you don't describe the problem you are having, I can't help. If it's about pip, you need to add the packages to the manifest. And flatpak-builder-tools can still help with it.

**Comment by lpyparmentier** [comment]
> > My answer to this comment https://github.com/flathub/flathub/pull/2648/files#r752235790
> 
> I don't see anything.
> 
> Anyway, there is no network access during build, but since you don't describe the problem you are having, I can't help. If it's about pip, you need to add the packages to the manifest. And flatpak-builder-tools can still help with it.

Got the following error if not present:

```
========================================================================
Building module streamdeck-ui in /home/lparmentier/projects/gpl/flathub/.flatpak-builder/build/streamdeck-ui-3
========================================================================
Note: switching to 'v1.1.2'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at f94f95e Fix regression in key codes
Running: pip3 install --prefix=/app .
Processing /run/build/streamdeck-ui
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /tmp/pip-standalone-pip-kz2trfwj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-mzdj9x1d/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'poetry>=0.12'
       cwd: None
  Complete output (7 lines):
  WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f8145d09910>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/poetry/
  WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f8145d09e50>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/poetry/
  WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f8145d09a90>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/poetry/
  WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f8145d09d00>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/poetry/
  WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f8145d09220>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution')': /simple/poetry/
  ERROR: Could not find a version that satisfies the requirement poetry>=0.12 (from versions: none)
  ERROR: No matching distribution found for poetry>=0.12
  ----------------------------------------
WARNING: Discarding file:///run/build/streamdeck-ui. Command errored out with exit status 1: /usr/bin/python3 /tmp/pip-standalone-pip-kz2trfwj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-mzdj9x1d/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'poetry>=0.12' Check the logs for full command output.
ERROR: Command errored out with exit status 1: /usr/bin/python3 /tmp/pip-standalone-pip-kz2trfwj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-mzdj9x1d/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'poetry>=0.12' Check the logs for full command output.
Error: module streamdeck-ui: Child process exited with code 1
make: *** [Makefile:5: build] Error 1
```

I guess it should ideally solved by including a generated-pip.json or generated-poetry.json with [flatpak-builder-tools](https://github.com/flatpak/flatpak-builder-tools/tree/master/poetry), however I've been through and it was a nightmare (downloading pywin32 package and other unneeded package that make the pip-gen script fail and incomplete). 

**It is a no go for me if I should go through this again** [partial discussion](https://github.com/flatpak/flatpak-builder-tools/issues/239#issuecomment-971430197). I had to custom the script to disable the download/specification of some package (followed this [idea](https://github.com/flatpak/flatpak-builder-tools/pull/217)) then I add to manually add a lot of different wheels (setuptools, wheel, and other stuffs) to make the pip install properly work.

There is absolutely no pleasure and no interest to go this way. It's is unmaintainable, if the author add new dependencies, or change some version of the dependecies, you have to go through this whole tricky and tedious process again.

**Comment by lpyparmentier** [comment]
Why there is no network access ? is that for this reason that we need to use the builder tool ?

**Comment by hfiguiere** [comment]
> Why there is no network access ? 

*security*

> is that for this reason that we need to use the builder tool ?

Exactly. It's even explained in the documentation.

**Comment by lpyparmentier** [comment]
> > Why there is no network access ?
> 
> _security_
> 
> > is that for this reason that we need to use the builder tool ?
> 
> Exactly. It's even explained in the documentation.

I'm a little bit confused. It should be already secure enough through the isolation of bubblewarp + configuration (socket, filesystem and so on), no ? But making the update process hazardous will not allow to easily add new features (at least for python apps). I'm too lazy to track the change of apps and keep records of a generated-poetry/pip.json according to the project. I feel like flatpak is great for app developed in C/C++, but concerning python+qt it is different.

**Comment by hfiguiere** [comment]
there is a lot of python apps in flathub.

**Comment by lpyparmentier** [comment]
> there is a lot of python apps on flathub.

I'll go through the poetry tool build again, make an issue and detail the problem

**Comment by hfiguiere** [comment]
I wanted to add, despite the lack of activity of this, that the app-id is likely incorrect.

It should be related to upstream.

**Comment by lpyparmentier** [comment]
thanks @hfiguiere, but following my remarks in https://github.com/flatpak/flatpak-builder-tools/issues/245 I will not go forward on this PR

**Pull request closed without merging**